### PR TITLE
feat: upgrade project to use `badger/v4` 

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
### Description

- Update `badger/v2` to `badger/v4` since it contains substantial improvements 
- `badger/v2` was released in March 2020 while `badger/v4` was released in Feb 2023
- Since `badger/v3` in the `Subscribe` method, in the last argument, now requires a slice of the type `pb.Match` instead of a byte slice 
- `badger/v4` release details: https://github.com/dgraph-io/badger/releases/tag/v4.0.1
- blocked by: https://github.com/ChainSafe/chaindb/pull/25
- Update to badger v4.1.0 (April 2023) release details: https://github.com/dgraph-io/badger/releases/tag/v4.1.0

### CODE REVIEW

File: database.go [hunk 1]
```diff
@@ -184,7 +185,7 @@ func (db *BadgerDB) ClearAll() error {
 }
 
 // Subscribe to watch for changes for the given prefixes
-func (db *BadgerDB) Subscribe(ctx context.Context, cb func(kv *KVList) error, prefixes []byte) error {
+func (db *BadgerDB) Subscribe(ctx context.Context, cb func(kv *KVList) error, prefixes []pb.Match) error {
```
Comment (timwu20) [inline database.go:188]: what is `pb.Match`?

Comment (EclesioMeloJunior) [reply database.go:188]: the type `pb.Match` which is a structure defined here: https://github.com/dgraph-io/badger/blob/31e5a1bd0e04a5a0583169d5298b61744f105996/pb/badgerpb3.pb.go#L501, and it holds a slice of bytes, so the main change is that from an unwrapped byte slice to a structure that holds it. This type was introduced by `badger/v3`

**Review by jimjbrettj — APPROVED** [review]
lgtm!

**Comment by dimartiro** [comment]
@EclesioMeloJunior Can we merge this PR and start using the chaindb with badger v4?

**Comment by dimartiro** [comment]
@EclesioMeloJunior Can we upgrade it to badger [4.1.0](https://github.com/dgraph-io/badger/releases/tag/v4.1.0)?

**Review by timwu20 — APPROVED** [review]

**Pull request merged**